### ISSUE TITLE
Promise chain friendly errors

### ISSUE THREAD

**Issue by xpepermint** [ansman/validate.js#51]
Hi. I started using promises all over my code because it makes the code look beautiful. I really like this library but the problem is that errors are returned as `Object` not as `instanceof` some `ValidationError` class.

What I would propose is to wrap errors into `validate.ValidationError` class where you can still get an object calling `toObject()` method thus everything works the same way we just add a better support for promises.

Example:

```js
  validate.async(req.body, contraints).then(function() {
    return Model.save(req.body);
  }).then(function(result) {
    res.status(201).json(result);
  }).catch(validate.ValidationError, function(err) {
    res.status(422).json(err);
  }).catch(function(err) {
    res.status(500).json(err);
  });
```

I would appreciate if we can do this small update. Thanks!

**Comment by ansman** [comment]
I had something like this before but it caused a bunch of problems (mostly because I extended the error with the result) but I suppose this feature makes sense and I do like clean code :)

I'll give it some thought for the next version.

**Comment by ansman** [comment]
This relates to #10 BTW.

Would the solution posted in that issue work for you?
```javascript
validate.async(req.body, contraints).then(function() {
    return Model.save(req.body);
  }).then(function(result) {
    res.status(201).json(result);
  }).catch(Error, function(err) {
    res.status(500).json(err);
  }).catch(function(err) {
    res.status(422).json(err);
  });
```

**Comment by xpepermint** [comment]
Yeah, I saw that example. The problem is that usually the last `catch` is a system (unhandled) error which should not supposed to happen and must have a status 50X. I handle the last `catch` globally and a user gets only a simple error message, the rest is sent to a developer. 

I really feel like the validator should return an instance of a class thus the error could be distinguished in try/catch block (promise). Maybe we could add it as configuration option or something. 

The error class could look something like this:

```js
module.exports = function ValidationError(params, message, options) {
  Error.captureStackTrace(this, this.constructor);
  this.name = this.constructor.name;
  this.params = params;
  this.message = message || 'Validation failed';
  this.options = options;
};

require('util').inherits(module.exports, Error);
```

**Comment by ansman** [comment]
Yeah, agreed.
I think I always assumed that the error callback would be on the promise returned from validate.async but in your case that isn't true.

I'll probably add a `wrapAsyncErrors` function that will create the result object.

**Comment by xpepermint** [comment]
Sounds good. Thanks!

**Comment by xpepermint** [comment]
Hey @ansman. Any progress on this?

**Comment by xpepermint** [comment]
I created a simple upgrade so I can move on. Please check PR https://github.com/ansman/validate.js/pull/57.

**Comment by ansman** [comment]
Released in 0.8.0 (42e5cb3451dbd39d45a13e200fc6b3e199c69b77)
